[Reland] Added reference tests to ReductionOpInfo

heitorschueroff opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #64273

Reintroduced sample_inputs_prod and constrained the range of values for large reference tests.

This reverts commit e4fd2ab59ce8645f5ae9477c7724b6af82124b3b.

Differential Revision: [D30672097](https://our.internmc.facebook.com/intern/diff/D30672097)

heitorschueroff committed:
[Reland] Added reference tests to ReductionOpInfo

This reverts commit e4fd2ab59ce8645f5ae9477c7724b6af82124b3b.

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30672097).

heitorschueroff committed:
Update on "[Reland] Added reference tests to ReductionOpInfo"


Reintroduced sample_inputs_prod and constrained the range of values for large reference tests.

This reverts commit e4fd2ab59ce8645f5ae9477c7724b6af82124b3b.

Differential Revision: [D30672097](https://our.internmc.facebook.com/intern/diff/D30672097)

[ghstack-poisoned]

albanD commented on aten/src/ATen/native/cpu/ReduceOpsKernel.cpp:
```diff
@@ -163,24 +163,29 @@ static void std_var_kernel_impl(TensorIterator& iter, int64_t correction, bool t
-        [=](scalar_t a, scalar_t b) -> scalar_t { return a * b; },
-        [=](Vectorized <scalar_t> a, Vectorized <scalar_t> b) { return a * b; },
+        [=](scalar_t a, scalar_t b)
+            __ubsan_ignore_undefined__ -> scalar_t { return a && b; },
```
What does these guards do?
Why were they not needed before?

heitorschueroff replied:
They suppress undefined behavior error from https://clang.llvm.org/docs/UndefinedBehaviorSanitizer.html. These were needed because one of the builds was reporting undefined behavior on this kernel. This is used in other kernels as well.

albanD replied:
Is that because some of the new tests that are added?
And is it fine to ignore these?

heitorschueroff replied:
I assume it's because of the new tests since that's all that changed. I think it's safe to ignore it, we have other kernels doing the same. Ideally we should look into why the undefined behavior is happening and fix it though.

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -782,8 +784,8 @@ def _generate_reduction_inputs(device, dtype, requires_grad):
     yield make_tensor([2], device, dtype, requires_grad=requires_grad)
-    yield make_tensor([2, 3], device, dtype, requires_grad=requires_grad, noncontiguous=True)
-    yield make_tensor([3, 2, 1, 5], device, dtype, requires_grad=requires_grad)
+    yield make_tensor([3, 5], device, dtype, requires_grad=requires_grad, noncontiguous=True)
```
Why the change of size?

heitorschueroff replied:
To keep the test cases small and prevent overflow with prod.

albanD replied:
You go from 2, 3 to 3, 5. That doesn't really make the test case smaller? :D

heitorschueroff replied:
I made the [3, 2, 1, 5] -> [3, 2, 1, 2] smaller and [2, 3] -> [3, 5] bigger to compensate? Haha, these are arbitrary numbers at the end of the day.

albanD replied:
Yes they are. That's why I was surprised you changed them.

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -9288,6 +9336,37 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+            }), 'TestReductions', 'test_ref_small_input'),
+        ),
+        skips=(
+            # FIXME: prod does not support passing keepdim without passing dim
```
These comments should mention mean instead of prod?

heitorschueroff replied:
Addressed

mruberry commented:
fyi paralleltbb failure is unrelated and now resolved

mruberry commented:
@albanD you're shepherding this, right?

albanD commented:
@mruberry yes I can take a look.

heitorschueroff committed:
Update on "[Reland] Added reference tests to ReductionOpInfo"


Reintroduced sample_inputs_prod and constrained the range of values for large reference tests.

This reverts commit e4fd2ab59ce8645f5ae9477c7724b6af82124b3b.

Differential Revision: [D30672097](https://our.internmc.facebook.com/intern/diff/D30672097)

[ghstack-poisoned]

heitorschueroff commented:
@albanD This should be ready for another review

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30672097).

heitorschueroff commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -5720,6 +5723,53 @@ def gradcheck_wrapper_triangular_input(op, input, *args, upper=False, **kwargs):
+            if x.ndim == 0:
+                # NumPy reductions don't accept dim=0 for scalar inputs
+                for i in dim if isinstance(dim, tuple) else (dim,):
+                    assert i in {0, -1}
```
Changed the logic so the error is delegated to the actual operation.

heitorschueroff committed:
Update on "[Reland] Added reference tests to ReductionOpInfo"


Reintroduced sample_inputs_prod and constrained the range of values for large reference tests.

This reverts commit e4fd2ab59ce8645f5ae9477c7724b6af82124b3b.

Differential Revision: [D30672097](https://our.internmc.facebook.com/intern/diff/D30672097)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30672097).

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -188,6 +188,8 @@ def _np(t):
             elif isinstance(x, dict):
                 return {k: to_numpy(v) for k, v in x.items()}
+            elif isinstance(x, torch.dtype):
+                return torch.empty(0, dtype=x).numpy().dtype
```
Interesting move :D
Maybe we can re-use: https://github.com/pytorch/pytorch/blob/421d8f86b6def536df18371a5da2f5df4de6e262/torch/testing/_internal/common_utils.py#L455 ?

heitorschueroff replied:
> I'm not super happy with all the tolerance overrides but I think the PR is good :)

Could you elaborate more on your concerns about the tolerance overrides? Do you think we should skip the tests instead? Or if it a concern with the numerical instability they expose?

albanD replied:
If so many functions need to override the default value, maybe the default is not good?
Or we want to run these in double to make sure they are correct (like we do for gradcheck)?

heitorschueroff replied:
The problem is mostly with sum and prod for float16. I changed the decorator to override the tolerance for float16 for all tests. Some tests are specifically testing numerical stability so it's expected we need to override the tolerance, this is also a way to document numerical instability.

heitorschueroff replied:
I changed the large reference tests from float32 to float64

albanD reviewed (commented):
I'm not super happy with all the tolerance overrides but I think the PR is good :)

heitorschueroff committed:
Update on "[Reland] Added reference tests to ReductionOpInfo"


Reintroduced sample_inputs_prod and constrained the range of values for large reference tests.

This reverts commit e4fd2ab59ce8645f5ae9477c7724b6af82124b3b.

Differential Revision: [D30672097](https://our.internmc.facebook.com/intern/diff/D30672097)

[ghstack-poisoned]

heitorschueroff committed:
Update on "[Reland] Added reference tests to ReductionOpInfo"


Reintroduced sample_inputs_prod and constrained the range of values for large reference tests.

This reverts commit e4fd2ab59ce8645f5ae9477c7724b6af82124b3b.

Differential Revision: [D30672097](https://our.internmc.facebook.com/intern/diff/D30672097)

[ghstack-poisoned]

heitorschueroff commented:
@albanD What else is this PR missing to get accepted?

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30672097).

heitorschueroff committed:
Update on "[Reland] Added reference tests to ReductionOpInfo"


Reintroduced sample_inputs_prod and constrained the range of values for large reference tests.

This reverts commit e4fd2ab59ce8645f5ae9477c7724b6af82124b3b.

Differential Revision: [D30672097](https://our.internmc.facebook.com/intern/diff/D30672097)

[ghstack-poisoned]

heitorschueroff commented:
@heitorschueroff has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30672097).

albanD reviewed (approved):
Thanks for the update!
